File: backend/core/src/modules/ai/ai.service.ts
import { BadGatewayException, HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { InjectRepository } from '@nestjs/typeorm';
import { MoreThanOrEqual, Repository } from 'typeorm';
import { ProjectAccessService } from '../../common/access/project-access.service';
import { AiRequest, AiUsage } from '../../entities';

type RequestUser = { userId: string; role?: string };

@Injectable()
export class AiService {
  constructor(
    private readonly configService: ConfigService,
    @InjectRepository(AiRequest)
    private readonly requestsRepo: Repository<AiRequest>,
    @InjectRepository(AiUsage)
    private readonly usageRepo: Repository<AiUsage>,
    private readonly projectAccessService: ProjectAccessService,
  ) {}

  private getMonthlyLimit(role?: string): number | null {
    switch (role) {
      case 'admin':
      case 'master':
        return null;
      case 'pro':
        return 500;
      case 'user':
      default:
        return 50;
    }
  }

  private startOfCurrentMonthUtc(now = new Date()): Date {
    return new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 1, 0, 0, 0, 0));
  }

  async getQuota(userId: string, role?: string) {
    const limit = this.getMonthlyLimit(role);
    const used = await this.requestsRepo.count({
      where: {
        userId,
        createdAt: MoreThanOrEqual(this.startOfCurrentMonthUtc()),
      } as any,
    });

    return {
      limit,
      used,
      remaining: limit === null ? null : Math.max(limit - used, 0),
    };
  }

  private async assertQuota(userId: string, role?: string) {
    const { limit, used } = await this.getQuota(userId, role);
    if (limit !== null && used >= limit) {
      throw new HttpException('AI 사용 한도를 초과했습니다.', HttpStatus.TOO_MANY_REQUESTS);
    }
  }

  private extractPrompt(payload: any): string | undefined {
    if (!payload || typeof payload !== 'object') return undefined;
    const candidate = payload.prompt ?? payload.query ?? payload.text ?? payload.scenario ?? payload.character ?? undefined;
    if (typeof candidate !== 'string') return undefined;
    const trimmed = candidate.trim();
    if (!trimmed) return undefined;
    return trimmed.length > 20_000 ? `${trimmed.slice(0, 20_000)}…` : trimmed;
  }

  async proxy(user: RequestUser, feature: string, path: string, payload: any) {
    await this.assertQuota(user.userId, user.role);

    const projectId = payload?.projectId as string | undefined;
    if (projectId) {
      await this.projectAccessService.assertProjectAccess(user.userId, projectId);
    }

    const provider = typeof payload?.provider === 'string' ? payload.provider : undefined;
    const model = typeof payload?.model === 'string' ? payload.model : undefined;

    const created = this.requestsRepo.create({
      userId: user.userId,
      projectId,
      feature,
      provider,
      model,
      prompt: this.extractPrompt(payload),
      status: 'pending',
    } as Partial<AiRequest>);
    const request = await this.requestsRepo.save(created);

    const aiBaseUrl = this.configService.get<string>('ai.serviceUrl') ?? 'http://ai-service:8000';
    const url = `${aiBaseUrl.replace(/\/$/, '')}${path}`;

    const { projectId: _projectId, ...forward } = payload ?? {};

    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 60_000);
      const resp = await fetch(url, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(forward),
        signal: controller.signal,
      });
      clearTimeout(timeout);

      const text = await resp.text();
      const parsed = text ? JSON.parse(text) : {};

      if (!resp.ok) {
        await this.requestsRepo.update(
          { id: request.id } as any,
          {
            status: 'failed',
            result: { error: parsed?.detail ?? parsed?.message ?? text } as any,
            completedAt: new Date(),
          } as any,
        );
        throw new BadGatewayException('AI 서비스 요청에 실패했습니다.');
      }

      await Promise.all([
        this.requestsRepo.update(
          { id: request.id } as any,
          {
            status: 'completed',
            result: parsed as any,
            completedAt: new Date(),
          } as any,
        ),
        this.usageRepo.save(
          this.usageRepo.create({
            userId: user.userId,
            feature,
            tokensUsed: 1,
            model,
            provider,
          } as any),
        ),
      ]);

      return parsed;
    } catch (error) {
      if (error instanceof BadGatewayException) {
        throw error;
      }
      await this.requestsRepo.update(
        { id: request.id } as any,
        {
          status: 'failed',
          result: { error: '요청 처리 중 오류가 발생했습니다.' } as any,
          completedAt: new Date(),
        } as any,
      );
      throw new BadGatewayException('AI 서비스 요청에 실패했습니다.');
    }
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/service/AiService.java
package store.pjcloud.cowrite.core.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.Map;
import java.util.UUID;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.server.ResponseStatusException;
import store.pjcloud.cowrite.core.config.AiProperties;
import store.pjcloud.cowrite.core.entity.AiRequest;
import store.pjcloud.cowrite.core.entity.AiUsage;
import store.pjcloud.cowrite.core.repository.AiRequestRepository;
import store.pjcloud.cowrite.core.repository.AiUsageRepository;

@Service
public class AiService {
    private final AiProperties aiProperties;
    private final AiRequestRepository requestsRepo;
    private final AiUsageRepository usageRepo;
    private final ProjectAccessService projectAccessService;
    private final ObjectMapper objectMapper;
    private final RestTemplate restTemplate = new RestTemplate();

    public AiService(AiProperties aiProperties,
                     AiRequestRepository requestsRepo,
                     AiUsageRepository usageRepo,
                     ProjectAccessService projectAccessService,
                     ObjectMapper objectMapper) {
        this.aiProperties = aiProperties;
        this.requestsRepo = requestsRepo;
        this.usageRepo = usageRepo;
        this.projectAccessService = projectAccessService;
        this.objectMapper = objectMapper;
    }

    private Integer getMonthlyLimit(String role) {
        if (role == null) return 50;
        return switch (role) {
            case "admin", "master" -> null;
            case "pro" -> 500;
            default -> 50;
        };
    }

    private OffsetDateTime startOfCurrentMonthUtc() {
        OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
        return OffsetDateTime.of(now.getYear(), now.getMonthValue(), 1, 0, 0, 0, 0, ZoneOffset.UTC);
    }

    public Map<String, Object> getQuota(UUID userId, String role) {
        Integer limit = getMonthlyLimit(role);
        long used = requestsRepo.countByUserIdAndCreatedAtAfter(userId, startOfCurrentMonthUtc());
        Integer remaining = limit == null ? null : Math.max(limit - (int) used, 0);
        Map<String, Object> quota = new java.util.HashMap<>();
        quota.put("limit", limit);
        quota.put("used", used);
        quota.put("remaining", remaining);
        return quota;
    }

    private void assertQuota(UUID userId, String role) {
        Map<String, Object> quota = getQuota(userId, role);
        Integer limit = (Integer) quota.get("limit");
        long used = (long) quota.get("used");
        if (limit != null && used >= limit) {
            throw new ResponseStatusException(HttpStatus.TOO_MANY_REQUESTS, "AI 사용 한도를 초과했습니다.");
        }
    }

    private String extractPrompt(Object payload) {
        if (!(payload instanceof Map<?, ?> map)) return null;
        Object candidate = map.get("prompt");
        if (candidate == null) candidate = map.get("query");
        if (candidate == null) candidate = map.get("text");
        if (candidate == null) candidate = map.get("scenario");
        if (candidate == null) candidate = map.get("character");
        if (!(candidate instanceof String)) return null;
        String trimmed = ((String) candidate).trim();
        if (trimmed.isBlank()) return null;
        return trimmed.length() > 20_000 ? trimmed.substring(0, 20_000) + "…" : trimmed;
    }

    public Object proxy(UserContext user, String feature, String path, Map<String, Object> payload) {
        assertQuota(user.userId(), user.role());

        Object projectIdObj = payload.get("projectId");
        if (projectIdObj instanceof String projectIdStr && !projectIdStr.isBlank()) {
            projectAccessService.assertProjectAccess(user.userId(), UUID.fromString(projectIdStr));
        }

        String provider = payload.get("provider") instanceof String s ? s : null;
        String model = payload.get("model") instanceof String s ? s : null;

        AiRequest request = new AiRequest();
        request.setUserId(user.userId());
        if (projectIdObj instanceof String pid && !pid.isBlank()) {
            request.setProjectId(UUID.fromString(pid));
        }
        request.setFeature(feature);
        request.setProvider(provider);
        request.setModel(model);
        request.setPrompt(extractPrompt(payload));
        request.setStatus("pending");
        request = requestsRepo.save(request);

        String base = aiProperties.getServiceUrl() == null ? "http://ai-service:8000" : aiProperties.getServiceUrl();
        String url = base.replaceAll("/$", "") + path;

        payload.remove("projectId");

        try {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_JSON);
            HttpEntity<Map<String, Object>> entity = new HttpEntity<>(payload, headers);
            ResponseEntity<String> resp = restTemplate.postForEntity(url, entity, String.class);

            if (!resp.getStatusCode().is2xxSuccessful()) {
                requestsRepo.save(updateRequestFailed(request, resp.getBody()));
                throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "AI 서비스 요청에 실패했습니다.");
            }

            Object parsed = resp.getBody() == null ? Map.of() : objectMapper.readValue(resp.getBody(), Object.class);
            request.setStatus("completed");
            request.setResult(objectMapper.convertValue(parsed, Map.class));
            request.setCompletedAt(OffsetDateTime.now());
            requestsRepo.save(request);

            AiUsage usage = new AiUsage();
            usage.setUserId(user.userId());
            usage.setFeature(feature);
            usage.setTokensUsed(1);
            usage.setModel(model);
            usage.setProvider(provider);
            usageRepo.save(usage);

            return parsed;
        } catch (Exception ex) {
            requestsRepo.save(updateRequestFailed(request, "요청 처리 중 오류가 발생했습니다."));
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "AI 서비스 요청에 실패했습니다.");
        }
    }

    private AiRequest updateRequestFailed(AiRequest request, String error) {
        request.setStatus("failed");
        Map<String, Object> result = new java.util.HashMap<>();
        result.put("error", error == null ? "AI 서비스 요청에 실패했습니다." : error);
        request.setResult(result);
        request.setCompletedAt(OffsetDateTime.now());
        return request;
    }

    public record UserContext(UUID userId, String role) {}
}
